### ISSUE TITLE
Should all Transformers be passing **kwargs to superclass constructor?

### ISSUE THREAD

**Issue by dwf** [mila-udem/fuel#134]
A whole bunch of Transformers don't pass on extra arguments to their superclasses. It also strikes me as a little weird that they all end up with a `batch_input` attribute that in many cases is ineffectual.

**Comment by vdumoulin** [comment]
We should do a cleanup of the whole transformers module. I'm not a big fan of the distinction between examplewise and batchwise processing, as it introduces unnecessary complications in the implementation side in my opinion.

**Comment by dwf** [comment]
@bartvm mentioned a refactoring plan that might have the potential to keep
everybody happy, but I honestly forget what it was, I will dig up my IM
logs if he hasn't chimed in by the time I do.

**Comment by dwf** [comment]
Another, related issue, before I forget, is whether transformers are
allowed to either reuse numpy arrays they return. Time spent allocating
large-ish temporaries really adds up, not to mention the memory overhead.

**Comment by dwf** [comment]
@vdumoulin and I talked about it offline and I guess we are both of the opinion that the examples vs. batches thing adds unnecessary complication (multiple code paths when, in many cases, there could be one). It mirrors the situation in Blocks where VariableFilter could take *either* a role *or* a list of roles -- we went with "list in all cases" as a simplification. Alas we'd prefer to get rid of it, but I'm pretty sure that is a non-starter for @bartvm and/or @rizar (and maybe other stakeholders like Jan Chorowski or someone). I do think we could probably do something in-between and have the protocol always pass around batches but have a decorator that let you write example-wise code, did the unpacking and repacking logic, and raised an error if the batch size was greater than 1.

**Comment by rizar** [comment]
I guess that the only motivation for it is performance. That is, when ones whitens the data, it is a matrix multiplication, and a matrix multiplication is more effective batch-wise.

I will refresh my knowledge on how it is actually implemented and put my 2 cents.

**Comment by rizar** [comment]
It seems to me that we should keep it possible to introduce an optimized batch-wise computation pathway. On the other hand, this feature should be transparent to a user that does not need it. And yeah, some transformers, like `Padding`, can work on batches only.

I like your suggestion about a decorator, @dwf, but there might be cases when examples are packed in a batch in a non-trivial way, i.e. not the first dimension stands for the index in the batch. Still, I buy it, since can override the default behavior in such rare cases.

Another thing is that `batch_input` is seems rather redundant to me, because a transformer could ask the data stream it wraps what it produces. And the top data stream, which is a `DataStream` can ask this question to the iteration scheme. 

On a related note: I don't like that most of transformers accept a `request` parameter and simply check that it's `None`. 

To sum it up, a refactoring plan could look like this:

- all the transformers and also `DataStream` derive the type (batch vs examples) of their return value
- default `get_data` method works follows: it checks that no request is given, and 
  - if input consists of examples, it iterates over examples and calls a new `transform` method on each of them
  - if input consists of batches _and_ `transform_batch` method is available, it calls this method
  - otherwise it unpacks batches, calls `transform` and packs the batches back
- the awkwardly named `SingleMapping` gets renamed into `SourceWiseTransformer`, and instead of a custom `mapping` method, uses `transform` and `transform_batch`
- the data streams that do use `request` can keep overriding the `get_data` method

Do you like it, guys?

**Comment by vdumoulin** [comment]
@rizar I like your suggestion. I think we should clarify how examples and batches of examples should be organized. My understanding is the following:

* Both example and batch of examples are tuples, with one element per source (e.g. a data stream with sources `('features', 'targets')` will return a tuple with two elements, the first corresponding to features and the second corresponding to targets).
* Elements of the batch tuple are iterables (`list`, `tuple`, `numpy.ndarray`). In the case of lists and tuples, elements represent examples. In the case of numpy arrays, one of the axes is the batch axis.

Is it accurate?

**Comment by rizar** [comment]
Sorry for late reply! This format is described is the de facto standard, perhaps we can even fix it de jure. To clarify: the examples are officially tuples with one element per source, but the way batches are introduced has not yet been so far formalized.

**Comment by vdumoulin** [comment]
Thanks for the confirmation. In that case, your suggestion makes sense: whether tuple elements are single examples or batches of examples is arbitrary and should be resolved by asking the previous scheme/stream how to interpret its output.

I'll try to draft a PR that implements your idea so that we can discuss it in more concrete terms.
